### PR TITLE
Rules-Engine v2 - Task and Target Improvements

### PR DESCRIPTION
**Pull request by kennsippell**
# Description

This is an implementation of the "Tasks and Targets Improvements" as speced [here](https://docs.google.com/document/u/1/d/1sMrsop7mxURJg05sUsl9mqG3fvZcbzxxJGvU6ysRJaY/edit).

Includes: #5657, #5866, #5564

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/src/js/services/rules-engine.js [hunk 1]
```diff
@@ -1,256 +1,182 @@
...
+    
+    tasksOwnedBy: contactIds => fetchTasksFor(contactIds, 'owner'),
+
+    allContactDocs: function () {
+      return ContactTypes.getPersonTypes()
+      .then(function(personTypes) {
+        const keys = personTypes.map(personType => ([personType.id]));
+        return DB()
+          .query('medic-client/doc_by_type', { keys, include_docs: true })
+          .then(res => res.rows.map(row => row.doc).filter(doc => doc));
```
Comment (garethbowen) [inline webapp/src/js/services/rules-engine.js]: This won't work for configurable hierarchy as all contacts will have doc.type = `contact`. Consider using `contacts_by_type` instead.

File: ddocs/medic-client/views/tasks/map.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  if (doc.type === 'task') {
+    if (doc.owner) {
+      var isTerminalState = ['Cancelled', 'Completed', 'Failed'].indexOf(doc.state) >= 0;
+      if (!isTerminalState) {
+        emit('owner-' + doc.owner, { _id: doc.id });
+      }
+    }
+
+    if (doc.requester) {
+      emit('requester-' + doc.requester, { _id: doc.id });
```
Comment (garethbowen) [inline ddocs/medic-client/views/tasks/map.js]: You're emitting a value of `{ _id: doc.id }` which is bad for performance, and as far as I can tell unused.

**Comment by garethbowen** [comment]
Nice work! My general thoughts so far...

1. There's a lot of coupling between the RulesEngine and the shared-lib which makes it hard to follow. Ideally the shared-lib would only expose 2-3 functions to make the API really simple. Things like whether or not a contact is dirty should be an implementation detail and not exposed directly to the webapp. Is it possible to move most of the RulesEngine into the shared-lib so it's more cohesive?
2. I'm a bit confused about IDs. Is the emission ID what's emitted from the nools configuration? Is that stored as the `doc.id`? Can we name it better (`doc.id` and `doc._id` is confusing)?
3. For unit tests I'm not a fan of using mock objects - they frequently end up a mess as you have to keep extending them to handle more and more use cases. Can the MockRulesEmitter be replaced by sinon calls?

**Comment by kennsippell** [comment]
> 1. There's a lot of coupling between the RulesEngine and the shared-lib which makes it hard to follow. Ideally the shared-lib would only expose 2-3 functions to make the API really simple. Things like whether or not a contact is dirty should be an implementation detail and not exposed directly to the webapp. Is it possible to move most of the RulesEngine into the shared-lib so it's more cohesive?

As discussed offline, I'll move some of the RulesEngine logic into the rules-core. However, my goal of the shared-lib is that it remains independant of Pouch for re-use in the medic-conf-test-harness. So I'm unsure where to draw the line between our angular services and the pure business logic currently in the rules-core.

> 2. Im a bit confused about IDs. Is the emission ID what's emitted from the nools configuration? Is that stored as the doc.id? Can we name it better (doc.id and doc._id is confusing)?

An emission is the raw object passed from nools to the webapp. It has an id property which is the "emission ID".  A minified version of the emission is stored in the task document. This does include the emission's ID.  The task document's id is based around the emission id, but also includes the timestamp for when the doc was created (your suggestion). Do you have suggestions for how to make this clearer?

> 3. For unit tests I'm not a fan of using mock objects - they frequently end up a mess as you have to keep extending them to handle more and more use cases. Can the MockRulesEmitter be replaced by sinon calls?

This isn't a true mock, but is rather a version of a real RulesEmitter object which has a `.restore()` method which allows it to be initialized mutilple times without throwing. I struggled to implement this generically through sinon. If I rename it to `RestorableRulesEmitter` would that be clearer?

**Comment by garethbowen** [comment]
> Do you have suggestions for how to make this clearer?

Ok so I understood it correctly. I think it'll be clearer when the code is complete.

> I struggled to implement this generically and cleanly through sinon. If I rename it to RestorableRulesEmitter would that be clearer?

No, it's not about the name. Thanks for trying with sinon!

File: webapp/src/js/services/rules-engine.js [hunk 3]
```diff
@@ -1,256 +1,114 @@
...
+      key: 'mark-contacts-dirty',
+      filter: change => !!change.doc && (ContactTypes.includes(change.doc) || isReport(change.doc)),
+      callback: change => {
+        const subjectId = isReport(change.doc) ? registrationUtils.getPatientId(change.doc) : change.id;
+        RulesEngineCore.updateTasksFor(DB(), subjectId);
+      },
+    });
+
+    const userLineage = [];
+    for (let current = userContactDoc, i = 0; !!current && i < 50; current = current.parent, i++) {
```
Comment (garethbowen) [inline webapp/src/js/services/rules-engine.js]: You don't need `i` because you already know `userLineage.length`. Consider:

```js
for (let current = userContactDoc; !!current && userLineage.length < 50; current = current.parent) {
```

Also, "50" is a magic number so should be a constant not just inlined into a for loop.

File: webapp/src/js/services/target-generator.js [hunk 4]
```diff
@@ -1,121 +1,16 @@
...
-                  _targets.forEach(mergeTarget);
-                }
-                callback(err, targets);
-              });
-            })
-            .catch(callback);
-        } else {
-          callback(null, []);
-        }
+        callback(null, []);
```
Comment (garethbowen) [inline webapp/src/js/services/target-generator.js]: A big ol' TODO

File: config/default/translations/messages-en.properties [hunk 5]
```diff
@@ -119,7 +119,7 @@ targets.anc.facility_deliveries.title = In-facility deliveries
 targets.anc.new_pregnancy_registrations.title = New pregnancies
 targets.anc.newborn_mortality.title = Newborn mortality
 targets.death_reporting.deaths.title = Deaths
-targets.disabled = Targets are disabled for admin users.  If you need to see targets, login as a normal user.
+targets.disabled = Targets are disabled for your user account.
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply config/default/translations/messages-en.properties]: Oops. Edited wrong file. Moved change to defaults.

File: shared-libs/rules-engine/src/contact-state-store.js [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+  /**
+   * @returns {string} User contact document
+   */
+  currentUser: () => currentUser,
+};
+
+const hashRulesConfig = (settingsDoc, userDoc, salt) => {
+  const rulesConfig = {
+    version: '1', // change this release-to-release to force task recalculation after deployment
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/contact-state-store.js]: I thought about using the app version, but I don't think it is necessary to force the recalculation of tasks for every version change. For example, 3.7.1 doesn't need to force a recalculation of tasks. I'll just remove the line for now.

File: shared-libs/rules-engine/src/index.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+   * @param {Object} userDoc User's hydrated contact document
+   * @param {Object} salt=1 Salt to add into the configuration hash. Changing this value invalidates the cache.
+   */
+  rulesConfigChange: (settingsDoc, userDoc, salt = 1) => {
+    const cacheIsReset = contactStateStore.rulesConfigChange(settingsDoc, userDoc, salt);
+    if (cacheIsReset) {
+      rulesEmitter.shutdown();
+    }
+  },
+};
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/index.js]: Given `module.exports = { myInterface: lib.otherInterface };`, the jsDocs from `otherInterface` will carry over so there is no need to duplicate that documentation. Added docs for isEnabled.

File: shared-libs/rules-engine/src/task-fetcher.js [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        docsOf(db.query('medic-client/contacts_by_type', { include_docs: true })),
+        docsOf(db.query('medic-client/reports_by_subject', { include_docs: true })),
+        fetch.allTasks(db, 'requester'),
+      ])
+      .then(([contactDocs, reportDocs, taskDocs]) => ({ contactDocs, reportDocs, taskDocs, userContactId }));
+  },
+
+  /*
+  PouchDB.query slows down when provided with a large keys array.
+  For users with ~1000 users it is ~50x faster to provider a start/end key instead of specifying all ids
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/task-fetcher.js]: ~1000 contacts. good catch

File: shared-libs/rules-engine/src/task-fetcher.js [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+// previousCommit helps avoid conflict errors since these changes are handled asynchronously
+let previousCommit = Promise.resolve();
+const stateChangeCallback = (db, stateDoc, updatedState) => {
+  Object.assign(stateDoc, { contactStateStore: updatedState });
+
+  previousCommit = previousCommit
+    .then(() => db.put(stateDoc))
+    .then(updatedDoc => { stateDoc._rev = updatedDoc.rev; })
+    .catch(err => console.error(`Error updating contactStateStore: ${err}`))
+    .then(() => { previousCommit = Promise.resolve(); });
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/task-fetcher.js]: I added it because I wasn't sure of the memory implications of having a very long promise change.  Although previousCommit does resolve, if we keep adding promises to the chain could this bloat?

File: shared-libs/rules-engine/src/refresh-tasks-for-contacts.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+ * @param {Object[]} freshData.taskDocs All of the contacts' task documents (must be linked by requester to a contact)
+ * @param {Object[]} freshData.userContactId The id of the user's contact document
+ * @param {int} calculationTimestamp Timestamp for the round of rules calculations
+ *
+ * @returns {Object[]} Array of updated task documents
+ */
+module.exports = (freshData = {}, calculationTimestamp = Date.now()) => calculateFreshTaskDocs(freshData, calculationTimestamp)
+  .then(freshResults => {
+    const freshTaskDocs = freshResults.map(doc => doc.taskDoc);
+    const canceledDocs = getCancellationUpdates(freshTaskDocs, freshData.taskDocs, calculationTimestamp);
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/refresh-tasks-for-contacts.js]: Yep

File: shared-libs/rules-engine/src/transform-task-emission-to-doc.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  taskDoc.owner = taskEmission.contact && taskEmission.contact._id;
+  minifyEmission(taskDoc, taskEmission);
+  TaskStates.setStateOnTaskDoc(taskDoc, emittedState, calculatedAt);
+
+  const isUpdated =
+    // do not create new documents where the initial state is cancelled (invalid emission)
+    (baseFromExistingDoc || taskDoc.state !== TaskStates.Cancelled)
+    && (
+      !existingDoc ||
+      existingDoc._id !== taskDoc._id ||
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/transform-task-emission-to-doc.js]: I've reworked this into a funciton that is hopefully clearer.

File: shared-libs/rules-engine/src/transform-task-emission-to-doc.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      (
+        !TaskStates.isTerminal(existingDoc.state) &&  // reduce document churn - don't tweak data on docs in terminal states
+        JSON.stringify(taskDoc) !== JSON.stringify(existingDoc)
+      )
+    );
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/transform-task-emission-to-doc.js]: Done

File: shared-libs/rules-engine/src/transform-task-emission-to-doc.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+  return {
+    isUpdated,
+    taskDoc,
+  };
+};
+
+const minifyEmission = (taskDoc, emission) => {
+  const minified = _.pick(emission, '_id', 'title', 'icon', 'startTime', 'endTime', 'deleted', 'resolved', 'actions');
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/transform-task-emission-to-doc.js]: Little embarrassing after the script weight investigation.

File: webapp/package.json [hunk 14]
```diff
@@ -33,7 +33,6 @@
     "gsm": "^0.1.4",
     "jquery": "3.2.x",
     "lodash": "^4.17.15",
-    "medic-nootils": "^2.0.0",
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply webapp/package.json]: Missed it. Good catch.

File: webapp/src/js/controllers/tasks.js [hunk 15]
```diff
@@ -1,125 +1,168 @@
...
+      }
+      
+      return promiseToLoad.then(function () {
+        const task = LiveList.tasks.getList().find(taskEmission => taskEmission. _id === id);
         if (task) {
-          var refreshing = (ctrl.selectedTask && ctrl.selectedTask._id) === id;
+          const refreshing = (ctrl.selectedTask && ctrl.selectedTask._id) === id;
           ctrl.settingSelected(refreshing);
-          setSelectedTask(task);
+          return setSelectedTask(task);
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply webapp/src/js/controllers/tasks.js]: `setSelectedTask` now returns a promise. anything waiting for `setSelected` to complete will need to wait for `setSelectedTask` also

File: webapp/src/js/services/rules-engine.js (hunk 3 above)
Comment (kennsippell) [reply webapp/src/js/services/rules-engine.js]: Nice

File: webapp/src/js/services/target-generator.js (hunk 4 above)
Comment (kennsippell) [reply webapp/src/js/services/target-generator.js]: :)

File: webapp/src/js/app.js [hunk 16]
```diff
@@ -146,7 +146,7 @@ const createReduxLoggerConfig = Selectors => ({
     $compileProvider.aHrefSanitizationWhitelist(
       /^\s*(https?|ftp|mailto|tel|sms|file|blob):/
     );
-    var isDevelopment = window.location.hostname === 'localhost';
+    var isDevelopment = window.location.hostname === 'localhost' && window.location.port !== '9876'; // 9876 is port used for karma tests
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply webapp/src/js/app.js]: Currently a lot of our karma tests for controllers mock the redux actions/reducers. The tests for the tasks controller mock the redux store itself and confirm the data that is written to the store, not just the data that is passed to an action. It is a bit more of an integration test. I think this is a pattern we should encourage, particularly as we move more and more logic into redux. The redux logging that results for this style of test is quite verbose. It isn't a huge problem now, so happy to remove this line - but that was my thinking.

File: webapp/src/js/controllers/contacts.js [hunk 17]
```diff
@@ -454,13 +446,7 @@ const PAGE_SIZE = 50;
...
         return false;
       }
-      return Auth('can_view_last_visited_date')
-        .then(function() {
-          return true;
-        })
-        .catch(function() {
-          return false;
-        });
+      return Auth.has('can_view_last_visited_date');
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply webapp/src/js/controllers/contacts.js]: Reverted

File: shared-libs/rules-engine/src/contact-state-store.js [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    state = {
+      rulesConfigHash: hashRulesConfig(settingsDoc, userDoc),
+      contactStates: {},
+    };
+    currentUser = userDoc;
+
+    onStateChange = safeCallback(stateChangeCallback);
+    return onStateChange(state);
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/contact-state-store.js]: It actually is not undefined. The callback may return a promise, as [our concrete implementation does](https://github.com/medic/cht-core/pull/6041/files#diff-73cb3dd3f8d28408c50f89304011bf43R199). Passing this promise down to the caller gives the option to have code wait for the state change to commit before it proceeds to the next task. For example, we do that [here](https://github.com/medic/cht-core/pull/6041/files#diff-73cb3dd3f8d28408c50f89304011bf43R87) but choose not to [here](https://github.com/medic/cht-core/pull/6041/files#diff-73cb3dd3f8d28408c50f89304011bf43R118). 

If the state fails to commit, downstream code in the promise will abort.
This also makes integration testing a lot cleaner, since we can easily await the state change to commit before checking the content of the document.

I personally feel this is something we should keep, though I agree it may be possible to do without.

File: shared-libs/rules-engine/test/contact-state-store.spec.js [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+    const state = mockState({ 'a': { calculatedAt: Date.now() } });
+    state.rulesConfigHash = 'hash';
+    await contactStateStore.load(state, {}, {});
+
+    const isDirty = contactStateStore.isDirty('a');
+    expect(isDirty).to.be.true;
+  });
+
+  it('scenario after loading state', async () => {
+    const onStateChange = sinon.stub().resolves();
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/test/contact-state-store.spec.js]: See above

File: shared-libs/rules-engine/src/contact-state-store.js [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+  currentUser: () => currentUser,
+};
+
+const hashRulesConfig = (settingsDoc, userDoc, salt) => {
+  const rulesConfig = {
+    version: '1', // change this release-to-release to force task recalculation after deployment
+    rules: settingsDoc && settingsDoc.tasks && settingsDoc.tasks.rules,
+    targets: settingsDoc && settingsDoc.targets,
+    userDoc,
+    salt,
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/contact-state-store.js]: This is for the "recalculate rules" button that I was hoping to add beside the "debug mode" button on the about page. I think it would be quite useful for testing.

File: shared-libs/rules-engine/src/task-fetcher.js [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+          ])
+          .then(([reportDocs, taskDocs]) => ({
+            userContactId: userDoc && userDoc._id,
+            contactDocs,
+            reportDocs,
+            taskDocs,
+          }));
+      });
+  },
+};
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/task-fetcher.js]: Can you confirm you are proposing to move the `stateChangeCallback` function to a separate module?  Just wanting to clarify because I'm finding the line this comment is on to be slightly ambiguous.

File: shared-libs/rules-engine/src/contact-state-store.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply shared-libs/rules-engine/src/contact-state-store.js]: I think it would be clearer to have a `force` parameter which, if set, doesn't run the diff at all.

File: shared-libs/rules-engine/src/rules-emitter.js [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+  shutdown,
+};
+
+const startSession = function() {
+  if (!flow) {
+    throw Error('Failed to start task session. Not initialized');
+  }
+
+  const session = flow.getSession();
+  const tasks = [], targets = [];
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply shared-libs/rules-engine/src/rules-emitter.js]: I think so, yes. Unfortunately it may require a fair bit of work to get existing code to pass...

File: shared-libs/rules-engine/src/task-fetcher.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply shared-libs/rules-engine/src/task-fetcher.js]: I was thinking the whole `const fetch = {...}` object. It appears to be a well encapsulated inner object that could easily be extracted. I may be missing something though...

File: webapp/src/js/app.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/app.js]: I think we should keep a clear distinction between unit tests and integration tests, and therefore it would be more correct to mock the redux actions, and then write another unit test for redux which ensures the data is written to the store when the action is called.

As well as that you may want to write integration tests, but they serve a different purpose, and I don't want to confuse the issue by having some unit tests actually test _some_ integrations because it makes it unclear what our actual test coverage is.

The integration tests I really care about are end-to-end ones. For example, you're testing that the store is updated but not that the UI is updated when the store is written to. This would be relatively easy to write and test the integration from start to finish.

File: shared-libs/rules-engine/src/transform-task-emission-to-doc.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+
+ * @returns {Object} result
+ * @returns {Object} result.taskDoc The result of the transformation
+ * @returns {Boolean} result.isUpdated True if the document is new or has been altered
+ */
+module.exports = (taskEmission, calculatedAt, userContactId, existingDoc) => {
+  let emittedState = TaskStates.calculateState(taskEmission, calculatedAt);
+  const baseFromExistingDoc = !!existingDoc && (!TaskStates.isTerminal(existingDoc.state) || existingDoc.state === emittedState);
+
+  const taskDoc = baseFromExistingDoc ? deepCopy(existingDoc) : newTaskDoc(taskEmission, userContactId, calculatedAt);
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/transform-task-emission-to-doc.js]: I achieved some variation of this.

File: shared-libs/rules-engine/src/task-fetcher.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/task-fetcher.js]: ahhh okay. that makes more sense

File: webapp/src/js/app.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (kennsippell) [reply webapp/src/js/app.js]: although it is relatively easy to write a test that validates controller down to the UI, i find them cumbersome to maintain. they require xpaths to locations in the dom, which make them quite fragile.

> I think we should keep a clear distinction between unit tests and integration tests
where should i put integration tests then?

[the remark this answers is not in the document]
Comment (garethbowen) [reply webapp/src/js/app.js]: >  i'd write that in protractor instead of karma.

Definitely write e2e tests in protractor as part of the e2e suites we already have.

> where should i put integration tests?

In API we have a `tests/integration` directory. We could introduce the same thing in webapp?

File: shared-libs/rules-engine/src/contact-state-store.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/contact-state-store.js]: Just removed for now

File: shared-libs/rules-engine/src/task-fetcher.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (kennsippell) [reply shared-libs/rules-engine/src/task-fetcher.js]: i've completely decoupled task-fetcher from pouchdb

File: webapp/src/js/services/target-generator.js (hunk 4 above)
Comment (kennsippell) [reply webapp/src/js/services/target-generator.js]: https://github.com/medic/cht-core/pull/6114

**Comment by kennsippell** [comment]
My device for performance testing isn't working today, but I'll try to get that working. Here is an initial dataset from my laptop using the standard QA performance dataset.  I am performing an initial replication, then warming all views, then taking measures. To clear cache I run a script (below).

### Results
Measure | Baseline | Experiment | Delta
-- | -- | -- | --
Bootstrap | 3071 | 1479 | -1592 (-52%)
Tasks | 95 | 517 | +422 (+450%)
Targets | 130 | 134 | +4 (+3%)

Experiment after cache bust:
Tasks: 3378
Targets: 2727

Memory Consumption | (MB)
-- | --
Baseline | 85
Experiment | 129
Delta | +44 MB (51%)

### Clearing Cache
```
(async () => {
  const doc = await PouchDB('medic-user-ac1').get('_local/rulesStateStore');
  const updated = { _id: doc._id, _rev: doc._rev, rulesStateStore: {} };
  return await PouchDB('medic-user-ac1').put(updated);
})();
```
